crates/chronicle/src/codegen/mod.rs

```diff
@@ -774,6 +774,7 @@ fn gen_query() -> rust::Tokens {
     impl Query {
 
     #[allow(clippy::too_many_arguments)]
+    #[graphql(guard = "chronicle_graphql::AuthGuard::new(chronicle_graphql::AccessType::QUERY, \"activityTimeline\")")]
```

Comment:
I think we need a more generalised design than this, guards also need to be applied to resolvers as well as methods. Context has everything we need I think:

https://github.com/async-graphql/async-graphql/blob/master/src/context.rs#L250

If we generalise guard to use context internally, then it can also be easily applied to objects, with this approach that's going to be painful

Reply:
https://github.com/async-graphql/async-graphql/blob/e839ffafcce3b114289d61a54cb5edbe1377e1a0/tests/guard.rs#L58

See here, for what we need to do - all fields need to be guarded for fine grained query authorisation, not just the root query / mutations

Reply:
https://github.com/async-graphql/async-graphql/blob/ad05baf79aea55a0827291abf55af5ce7cc8d386/src/validation/rules/known_directives.rs#L25

```
     self.location_stack
            .push(match &operation_definition.node.ty {
                OperationType::Query => __DirectiveLocation::QUERY,
                OperationType::Mutation => __DirectiveLocation::MUTATION,
                OperationType::Subscription => __DirectiveLocation::SUBSCRIPTION,
            });
            ```

Context has everything we need to make an auth choice already parsed and available, there's no need to pass anything into the guard itself, we just need to get some things from the context and forward it to the opa executor along with the identity

Reply:
Thank you for these tips. My current impression is that we can get rid of those guards and instead hook into `async_graphql::extensions::Extension::resolve`. I shall experiment accordingly.

Reply:
@ryanbtp: It turns out that using another extension does appear to allow us to avoid touching codegen. The `resolve_path` that I use in the new rego file includes the specific elements that the request tries to query out of the server.